lib/phoenix/pubsub/redis_server.ex

```diff
@@ -0,0 +1,176 @@
+  is started successfully, it handles reconnections automatically, so we
+  pass off reconnection handling once we find an initial connection.
+  """
+  def init(opts) do
+    server_name = Keyword.fetch!(opts, :name)
+    local_name  = Keyword.fetch!(opts, :local_name)
+    opts = Dict.merge(@defaults, opts)
+    opts = Dict.merge(opts, host: String.to_char_list(to_string(opts[:host])),
+                            password: String.to_char_list(to_string(opts[:password])))
+
+    Process.flag(:trap_exit, true)
+    send(self, :establish_conn)
+
+    {:ok, %{local_name: local_name,
+            namespace: redis_namespace(server_name),
+            eredis_sub_pid: nil,
+            eredis_pid: nil,
+            status: :disconnected,
+            node_ref: :erlang.make_ref,
+            reconnect_attemps: 0,
+            opts: opts}}
+  end
+
+  def handle_call({:subscribe, pid, topic, link}, _from, state) do
+    if link, do: Process.link(pid)
+    {:reply, GenServer.call(state.local_name, {:subscribe, pid, topic}), state}
+  end
+
+  def handle_call({:unsubscribe, pid, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_name, {:unsubscribe, pid, topic}), state}
+  end
+
+  def handle_call({:subscribers, topic}, _from, state) do
+    {:reply, GenServer.call(state.local_name, {:subscribers, topic}), state}
+  end
+
+  def handle_call({:broadcast, from_pid, topic, msg}, _from, state) do
+    redis_msg = {@redis_msg_vsn, state.node_ref, from_pid, topic, msg}
+    case :eredis.q(state.eredis_pid, ["PUBLISH", state.namespace, redis_msg]) do
+      {:ok, _}         -> {:reply, :ok, state}
+      {:error, reason} -> {:reply, {:error, reason}, state}
+    end
+  end
+
+  @doc """
+  Decodes binary message and fowards to local subscribers of topic
+  """
+  def handle_info({:message, _redis_topic, bin_msg, _cli_pid}, state) do
+    {_vsn, remote_node_ref, from_pid, topic, msg} = :erlang.binary_to_term(bin_msg)
+
+    if remote_node_ref == state.node_ref do
+      GenServer.call(state.local_name, {:broadcast, from_pid, topic, msg})
+    else
+      GenServer.call(state.local_name, {:broadcast, :none, topic, msg})
+    end
+
+    :eredis_sub.ack_message(state.eredis_sub_pid)
+    {:noreply, state}
+  end
+
+  def handle_info({:EXIT, _pid, {:connection_error, {:connection_error, :econnrefused}}}, state) do
+    {:noreply, state}
+  end
+
+  def handle_info({:EXIT, _linked_pid, _reason}, state) do
+    {:noreply, state}
+  end
+
+  @doc """
+  Connection establishment and shutdown loop
+
+  On init, an initial conection to redis is attempted when starting `:eredis_sub`.
+  If failed, the connection is tried again in `@reconnect_after_ms` until a max
+  of `@max_connect_attemps` is tried, at which point the server terminates with
+  an `:exceeded_max_conn_attempts`.
+  """
+  def handle_info(:establish_conn, %{reconnect_attemps: count} = state)
+    when count >= @max_connect_attemps do
+
+    {:stop, :exceeded_max_conn_attempts, state}
```

Comment:
We still need to discuss if we are getting rid of this before merging as per our previous discussion about failure handling!